### PR TITLE
WrittenAction implementation

### PR DESCRIPTION
**Pull request by jml**
Adds a `WrittenAction` implementation to parallel the `WrittenMessage` implementation. 

The focus is on being able to make something that represents an action that you can trust. Another way to express this is that I'm trying to provide a definition in Python of what an action actually is (rather than a way of writing out an action). This means there's a hefty emphasis on validation.

As yet does not provide the ability to assemble a `WrittenAction` from a stream of messages, nor to assemble nested `WrittenAction` objects. I can add those if you'd like, but I thought the patch was big enough as it stands to be a reviewable chunk. Also, if what I think an action is is wrong, then there's not much point going further.

The tests are all written with Hypothesis. I don't know if this is actually a good idea. Although it looks like I'm adding 13 tests, I'm actually adding 2600. Currently the test suite takes ~34s on my laptop.

I've also introduced a testtools dependency, because Itamar implied that would be OK. It makes a lot of the aggregate assertions nicer.

This branch incorporates changes from #200 and would be much easier to review if that were merged.

[Review on Reviewable](https://reviewable.io/reviews/clusterhq/eliot/202)

### CODE REVIEW

File: eliot/tests/test_action.py [hunk 1]
```diff
@@ -905,3 +928,220 @@ def test_to_string(self):
...
+    return message.task_level.parent().level.append(n)
+
+
+class WrittenActionTests(testtools.TestCase):
+    """
+    Tests for L{WrittenAction}.
+    """
+
+    @given(START_ACTION_MESSAGES)
+    def test_from_single_message(self, message):
```
Comment (itamarst) [inline eliot/tests/test_action.py]: In general, please make sure tests have docstrings.

File: eliot/_action.py [hunk 2]
```diff
@@ -8,12 +8,12 @@
...
 
 import threading
-from uuid import uuid4
+from uuid import UUID, uuid4
 from itertools import count
 from contextlib import contextmanager
 from warnings import warn
 
-from pyrsistent import PClass, pvector_field
+from pyrsistent import field, PClass, optional, pmap_field, pvector_field, m, v
```
Comment (itamarst) [inline eliot/_action.py]: `m` and `v` are really terrible names for an API. Use `m as to_map` or something?

File: eliot/_action.py [hunk 3]
```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
...
+
+    def __init__(self, action, message):
+        Exception.__init__(
+            self, 'Tried to add {} to {}, but {} is not a sibling of {}'.format(
+                message, action, message.task_level, action.task_level))
+
+
+class WrongActionType(Exception):
+    """
+    Tried to end a message with a different action_type to the beginning.
```
Comment (itamarst) [inline eliot/_action.py]: "than the beginning"?

File: eliot/_action.py [hunk 4]
```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
...
+        yet, then C{None}.
+
+    @ivar reason: The reason the action failed. If the action succeeded, or
+        hasn't finished yet, then C{None}.
+
+    @ivar _children: A L{pvector} of L{WrittenAction} and L{WrittenMessage}
+        that make up this action.
+    """
+
+    # Broad philosophy:
```
Comment (itamarst) [inline eliot/_action.py]: Maybe this should be in the docstring?

File: eliot/_action.py [hunk 5]
```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
...
+        there has been no end message.
+
+    @ivar exception: If the action failed, the name of the exception that was
+        raised to cause it to fail. If the action succeeded, or hasn't finished
+        yet, then C{None}.
+
+    @ivar reason: The reason the action failed. If the action succeeded, or
+        hasn't finished yet, then C{None}.
+
+    @ivar _children: A L{pvector} of L{WrittenAction} and L{WrittenMessage}
```
Comment (itamarst) [inline eliot/_action.py]: Doesn't match what's below.

File: eliot/_action.py [hunk 6]
```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
...
+    end_time = field(type=optional(float), mandatory=True)
+    # XXX: Actually a map to either WrittenMessage or WrittenAction, but
+    # pyrsistent doesn't support recursive data types:
+    # https://github.com/tobgu/pyrsistent/issues/48
+    _children = pmap_field(TaskLevel, object)
+    exception = field(type=optional(unicode), mandatory=True)
+    reason = field(type=optional(unicode), mandatory=True)
+
+    # XXX: Possible invariants:
+    # - start_time <= end_time
```
Comment (itamarst) [inline eliot/_action.py]: Sadly not something we can rely on being true.

File: eliot/_action.py [hunk 7]
```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
...
+    # - task_level = set(child.task_level.parent() for child in children)
+    # - task_uuid = set(child.task_uuid for child in children)
+    # - exception and reason set iff status == FAILED_STATUS
```
Comment (itamarst) [inline eliot/_action.py]: Given arbitrary action types can't be guaranteed.

File: eliot/_action.py [hunk 8]
```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
...
+    # XXX: Actually a map to either WrittenMessage or WrittenAction, but
+    # pyrsistent doesn't support recursive data types:
+    # https://github.com/tobgu/pyrsistent/issues/48
+    _children = pmap_field(TaskLevel, object)
+    exception = field(type=optional(unicode), mandatory=True)
+    reason = field(type=optional(unicode), mandatory=True)
+
+    # XXX: Possible invariants:
+    # - start_time <= end_time
+    # - task_level = set(child.task_level.parent() for child in children)
```
Comment (itamarst) [inline eliot/_action.py]: This and next one seem reasonable. Delete this comment, in any case?

File: eliot/_action.py [hunk 9]
```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
...
+            C{action_status}, or has one that is not C{SUCCEEDED_STATUS} or
+            C{FAILED_STATUS}.
+        @raise InvalidStartMessage: If C{start_message} does not have a
+            C{ACTION_STATUS_FIELD} of C{STARTED_STATUS}, or if it has a
+            C{task_level} indicating that it is not the first message of an
+            action.
+
+        @return: A new, completed C{WrittenAction}.
+
+        @return: A new C{WrittenAction}.
```
Comment (itamarst) [inline eliot/_action.py]: Only need one of these :grin:

File: eliot/_action.py [hunk 10]
```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
...
+        @raise WrongTaskLevel: If C{message} has a C{task_level} that means
+            it's not a direct child.
+
+        @return: A new C{WrittenAction}.
+        """
+        self._validate_message(message)
+        level = message.task_level
+        if self._children.get(level, message) != message:
+            raise DuplicateChild(self, message)
+        return self.set(_children=self._children.set(level, message))
```
Comment (itamarst) [inline eliot/_action.py]: I recommend `transform()`! It's the bestest.

Comment (itamarst) [inline eliot/_action.py]: `return self.transform(["_children", level], message)`

File: eliot/_action.py [hunk 11]
```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
...
+
+    def _end(self, end_message):
```
Comment (itamarst) [inline eliot/_action.py]: The problem with this particular mechanism is that it loses any additional fields the end message may have. Success messages may have additional fields, at least. And now that I think of it there's the same problem with any extra messages the start message has.

So perhaps both start and end message should be attributes of the `WrittenAction`, and some of the things that are currently attributes become properties?

File: eliot/_action.py [hunk 12]
```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
...
+        C{WrittenAction}s.
+
+        @param WrittenMessage start_message: A message that has
+            C{ACTION_STATUS_FIELD}, C{ACTION_TYPE_FIELD}, and a C{task_level}
+            that ends in C{1}.
+        @param children: An iterable of C{WrittenMessage} and C{WrittenAction}
+        @param WrittenMessage end_message: A message that has the same
+            C{action_type} as this action.
+
+        @raise WrongTask: If C{end_message} has a C{task_uuid} that differs
```
Comment (itamarst) [inline eliot/_action.py]: I wonder if having individual exceptions is sufficiently useful that it's not worth just simplifying things by relying on pyrsistent invariants.

**Comment by itamarst** [comment]
So taking a step back, the assumption is that some other chunk of code recognizes start messages, end messages and child messages? Might just go with simple invariant if that's the case on assumption that chunk of code will already be doing this verification.

This in general looks pretty good, in particular more robust than the variant in `eliot.testing`. So please address and re-submit; I haven't read the tests in detail since the class structure may be changing a little.

File: eliot/_action.py (hunk 3 above)
Comment (jml) [inline eliot/_action.py]: It's [a regional thing](http://www.oxforddictionaries.com/words/different-from-than-or-to). I find "different than" awkward, but happy to standardize on US English.

File: eliot/tests/test_action.py (hunk 1 above)
Comment (jml) [inline eliot/tests/test_action.py]: Docstrings added to all tests.

File: eliot/_action.py (hunk 2 above)
Comment (jml) [inline eliot/_action.py]: There we have a difference of opinion. Changed to use `pmap` and `pvector`, which are equivalent in the case of constructing empty containers.

File: eliot/_action.py (hunk 3 above)
Comment (jml) [inline eliot/_action.py]: Updated.

File: eliot/_action.py (hunk 4 above)
Comment (jml) [inline eliot/_action.py]: I've moved the general sense of it to the docstring. The more aspirational parts have been deleted.

File: eliot/_action.py (hunk 5 above)
Comment (jml) [inline eliot/_action.py]: Corrected.

File: eliot/_action.py (hunk 6 above)
Comment (jml) [inline eliot/_action.py]: Quite. I think it would be an interesting thing for a log display tool to highlight.

File: eliot/_action.py (hunk 7 above)
Comment (jml) [inline eliot/_action.py]: Line removed.

File: eliot/_action.py (hunk 8 above)
Comment (jml) [inline eliot/_action.py]: Comment removed.

File: eliot/_action.py (hunk 10 above)
Comment (jml) [inline eliot/_action.py]: Ooh, it's like lenses. Thanks for the tip!

Changed.

File: eliot/_action.py (hunk 11 above)
Comment (jml) [inline eliot/_action.py]: I've made this change, and I like it a lot. 

It has the advantage of making it possible to "roundtrip" the `WrittenAction`, which is a great property. It also makes the methods simpler, at the cost of lots of fluffy property clauses.

**Comment by jml** [comment]
> So taking a step back, the assumption is that some other chunk of code recognizes start messages, end messages and child messages? Might just go with simple invariant if that's the case on assumption that chunk of code will already be doing this verification.

I think that's a possibility. Is it OK if we leave it as-is with the exceptions for now, but keep this in mind as an option when we add the code that recognizes it?

**Comment by jml** [comment]
Thanks for the review, it was really useful. I think I'm ready for another round.

File: eliot/_action.py [hunk 13]
```diff
@@ -415,6 +427,293 @@ def __exit__(self, type, exception, traceback):
...
+
+class WrittenAction(PClass):
+    """
+    An Action that has been logged.
+
+    This class is intended to provide a definition within Eliot of what an
+    action actually is, and a means of constructing actions that are known to
+    be valid.
+
+    @ivar _children: A L{pmap} from L{TaskLevel} to the L{WrittenAction} and
```
Comment (itamarst) [inline eliot/_action.py]: Document start_message and end_message.

File: eliot/_action.py [hunk 14]
```diff
@@ -415,6 +427,293 @@ def __exit__(self, type, exception, traceback):
...
+    """
+
+    start_message = field(type=WrittenMessage, mandatory=True)
+    end_message = field(type=optional(WrittenMessage), mandatory=True)
+    # XXX: Actually a map to either WrittenMessage or WrittenAction, but
+    # pyrsistent doesn't support recursive data types:
+    # https://github.com/tobgu/pyrsistent/issues/48
+    _children = pmap_field(TaskLevel, object)
+
+    # XXX: I'm fairly convinced that this would be simpler and clearer if we
```
Comment (itamarst) [inline eliot/_action.py]: I don't understand this comment.

File: eliot/_action.py [hunk 15]
```diff
@@ -415,6 +427,293 @@ def __exit__(self, type, exception, traceback):
...
+        The reason the action failed. If the action succeeded, or hasn't finished
+        yet, then C{None}.
+        """
+        if self.end_message:
+            return self.end_message.contents.get(REASON_FIELD, None)
+
+    @property
+    def children(self):
+        """
+        The list of child messages and actions, excluding the start and end
```
Comment (itamarst) [inline eliot/_action.py]: Children have a particular order, so I suggest sorting by task_level (task_level of messages for messages, task level of start message for actions).

File: eliot/tests/test_action.py [hunk 16]
```diff
@@ -905,3 +928,297 @@ def test_to_string(self):
...
+        such a message, raise an error.
+
+        This test handles the case where the status field is not present.
+        """
+        assume(ACTION_STATUS_FIELD not in message_dict)
+        message = WrittenMessage.from_dict(message_dict)
+        self.assertRaises(
+            InvalidStartMessage, WrittenAction.from_messages, message)
+
+    @given(START_ACTION_MESSAGE_DICTS, status=one_of(just(FAILED_STATUS), just(SUCCEEDED_STATUS), text()))
```
Comment (itamarst) [inline eliot/tests/test_action.py]: Too-long line.

File: eliot/tests/test_action.py [hunk 17]
```diff
@@ -905,3 +928,297 @@ def test_to_string(self):
...
+            }))
+        self.assertRaises(
+            InvalidStatus,
+            WrittenAction.from_messages, start_message, end_message=end_message)
+
+    @given(START_ACTION_MESSAGES, lists(MESSAGE_DICTS))
+    def test_children(self, start_message, child_messages):
+        """
+        We can construct a L{WrittenAction} with child messages. These messages
+        can be either L{WrittenAction}s or L{WrittenMessage}s. They are
```
Comment (itamarst) [inline eliot/tests/test_action.py]: I don't see where this second is actually proved.

Comment (itamarst) [inline eliot/tests/test_action.py]: I.e. that WrittenAction is supported.

File: eliot/tests/test_action.py [hunk 18]
```diff
@@ -905,3 +928,297 @@ def test_to_string(self):
...
+            WrongTask,
+            WrittenAction.from_messages, start_message, v(message))
+
+    @given(START_ACTION_MESSAGES, MESSAGE_DICTS)
+    def test_wrong_task_level(self, start_message, child_message):
+        """
+        All child messages of an action must have a task level that is a direct
+        child of the action's task level.
+        """
+        assume(not start_message.task_level.is_sibling_of(TaskLevel(level=child_message[TASK_LEVEL_FIELD])))
```
Comment (itamarst) [inline eliot/tests/test_action.py]: Too long line.

**Comment by itamarst** [comment]
OK, address the above comments, fix the deprecation error emitted by hypothesis, then merge.

**Comment by itamarst** [comment]
Thanks!

File: eliot/_action.py (hunk 14 above)
Comment (jml) [inline eliot/_action.py]: Deleted. (If you're really curious, I can explain on chat)

File: eliot/_action.py (hunk 13 above)
Comment (jml) [inline eliot/_action.py]: Done.

File: eliot/tests/test_action.py (hunk 16 above)
Comment (jml) [inline eliot/tests/test_action.py]: Fixed

File: eliot/tests/test_action.py (hunk 18 above)
Comment (jml) [inline eliot/tests/test_action.py]: Fixed.

**Comment by jml** [comment]
All the comments have been addressed. I think https://github.com/ClusterHQ/eliot/commit/24db74bfbd72eecce336f59f0202330e83a913b7 is a big enough chunk of code that I'd welcome a code review.

**Pull request merged**